### DISCUSSION TITLE
Making FeatureNode transparent

### DISCUSSION THREAD

**Discussion by OlaRune** [pelicanmapping/osgearth#2511] [Support Q&A] [answered]
Hi!

I'm trying to highlight a FeatureNode by making all other FeatureNodes transparent,

All nodes use the following style:
```
Basic_Polygon_Style {
stroke: #00ff00;
stroke-width: 3px;
stroke-tessellation-size: 10km;
stroke-smooth: true;
fill: #00ff00;
fill-opacity: 0.15;
outline: true;
text-align: center-center;
text-color: #00ff00;
text-halo: #000000;
altitude-clamping: terrain;
render-depth-offset: true;
}
```

I'm using the following function:
```
void AnnotationHandler::highlightNode(osg::Node* node)
{
    for (auto layer : shapeLayers)
    {
        osg::Group*  group       = layer->getGroup();
        const size_t numChildren = group->getNumChildren();

        for (size_t i = 0; i < numChildren; ++i)
        {
            osgEarth::FeatureNode* featureNode =
                dynamic_cast<osgEarth::FeatureNode*>(group->getChild(i));
            if (featureNode)
            {
                // Check if this is the node that should be highlighted
                if (featureNode == node)
                {
                    continue;
                }

                // Make all other nodes transparent
                osgEarth::Style style = featureNode->getStyle();

                osgEarth::LineSymbol* lineSymbol =
                    style.getSymbol<osgEarth::LineSymbol>();
                if (lineSymbol)
                {
                    osgEarth::Color strokeColor = lineSymbol->stroke()->color();
                    strokeColor.a()             = 0.25;
                    lineSymbol->stroke()->color() = strokeColor;
                }

                osgEarth::PolygonSymbol* polygonSymbol =
                    style.getSymbol<osgEarth::PolygonSymbol>();
                if (polygonSymbol)
                {
                    osgEarth::Color fillColor = polygonSymbol->fill()->color();
                    fillColor.a()             = 0.25;
                    polygonSymbol->fill()->color() = fillColor;
                }

                osgEarth::TextSymbol* textSymbol =
                    style.getSymbol<osgEarth::TextSymbol>();
                if (textSymbol)
                {
                    osgEarth::Color textColor   = textSymbol->fill()->color();
                    textColor.a()               = 0.25;
                    textSymbol->fill()->color() = textColor;
                }

                featureNode->setStyle(style);
                featureNode->dirty(); // Force redraw
            }
        }
    }
}
```

However, this only affects the TextSymbol and not the outline (LineSymbol Stroke) or the fill (PolygonSymbol Fill). Why?

**Comment by gwaldron** [comment]
I was not able to reproduce the issue with some test code ([here](https://gist.github.com/gwaldron/43a311af3fbed6a2a9ccc55945fc0d11)).

Are your FeatureNodes under the MapNode in the scene graph? If not, make sure you have the GL_BLEND mode enabled in OSG.

(Also, this might be a good use case for a shader instead. See `osgearth_pick.cpp` for an example.)

**Reply by OlaRune** [reply, marked as answer]
Thanks! After I got it to work with your exact polygon using my highlight function, but not my own polygons I discovered that I had applied my Basic_Polygon_Style to the Feature, not the FeatureNode. So after removing the style completely from the Feature and instead applying it to the FeatureNode I got it to work. I was modifying the wrong style.

**Comment by OlaRune** [comment]
@gwaldron Another related question. I'm trying to also make a PlaceNode with an IconSymbol transparent. The IconSymbol is a png loaded from file, that is already transparent in some parts.

I'm getting the IconSymbol from the style, then using getImage() to get the underlying Image. I then modify the alpha channel of the data vector in the Image, but the Image is not updated.

Just to make sure I'm doing the correct thing I tried setting all pixels to black as instead, but my Icon on the map remains unchanged.

How do I get the new Image to show up?

```
osgEarth::PlaceNode* placeNode = dynamic_cast<osgEarth::PlaceNode*>(node);
if (placeNode)
{
    osgEarth::Style style = placeNode->getStyle();
    osgEarth::IconSymbol* iconSymbol = style.getSymbol<osgEarth::IconSymbol>();
    if (iconSymbol)
    {
        osg::Image* image = iconSymbol->getImage();
        int pixelFormat = image->getPixelFormat();
        if (pixelFormat == GL_RGBA)
        {
            int totalPixels = image->s() * image->t();
            unsigned char* data = image->data();
            for (int i = 0; i < totalPixels; i++)
            {
                // TEMP SET ALL TO BLACK
                data[i * 4] = 0;
                data[i * 4 + 1] = 0;
                data[i * 4 + 2] = 0;
            }
        }
    }
    iconSymbol->setImage(image);
    placeNode->setStyle(style);
    placeNode->dirty(); // Force redraw
}
```

**Reply by OlaRune** [reply]
I can't even change the style to a different style with a different icon and get that one to update by calling ->dirty(). Is it not possible to change style on a PlaceNode?

**Reply by gwaldron** [reply]
Not sure -- you might need to call setDynamic(true) on any annotations you are going to be changing.

But it seems like this is overkill. Why not just use a simple shader to change the alpha?
